File: src/main/java/net/Chidoziealways/everythingjapanese/commands/ModArgumentTypes.java
package net.Chidoziealways.everythingjapanese.commands;

import net.Chidoziealways.everythingjapanese.EverythingJapanese;
import net.Chidoziealways.everythingjapanese.commands.argumentTypes.jutsu.JutsuArgument;
import net.minecraft.commands.synchronization.ArgumentTypeInfo;
import net.minecraft.commands.synchronization.ArgumentTypeInfos;
import net.minecraft.commands.synchronization.SingletonArgumentInfo;
import net.minecraft.core.registries.Registries;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.RegistryObject;

public class ModArgumentTypes {
    public static final DeferredRegister<ArgumentTypeInfo<?, ?>> ARGUMENT_TYPES =
            DeferredRegister.create(Registries.COMMAND_ARGUMENT_TYPE, EverythingJapanese.MOD_ID);

    public static final RegistryObject<ArgumentTypeInfo<JutsuArgument, ?>> JUTSU_ARGUMENT =
            ARGUMENT_TYPES.register("jutsu", () -> ArgumentTypeInfos.registerByClass(JutsuArgument.class, SingletonArgumentInfo.contextAware(JutsuArgument::jutsu)));

    public static void register(IEventBus eventBus) {
        ARGUMENT_TYPES.register(eventBus);
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/commands/ModArgumentTypes.kt
package net.Chidoziealways.everythingjapanese.commands

import net.Chidoziealways.everythingjapanese.MOD_ID
import net.Chidoziealways.everythingjapanese.commands.argumentTypes.jutsu.JutsuArgument
import net.minecraft.commands.CommandBuildContext
import net.minecraft.commands.synchronization.ArgumentTypeInfo
import net.minecraft.commands.synchronization.ArgumentTypeInfos
import net.minecraft.commands.synchronization.SingletonArgumentInfo
import net.minecraft.core.registries.Registries
import net.neoforged.bus.api.IEventBus
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Supplier
import java.util.function.Function

object ModArgumentTypes {
    val ARGUMENT_TYPES = DeferredRegister.create<ArgumentTypeInfo<*, *>?>(Registries.COMMAND_ARGUMENT_TYPE, MOD_ID)

    val JUTSU_ARGUMENT =
        ARGUMENT_TYPES.register(
            "jutsu")
            { ->
                ArgumentTypeInfos.registerByClass(
                    JutsuArgument::class.java,
                    SingletonArgumentInfo.contextAware<JutsuArgument?>(Function { context: CommandBuildContext? ->
                        JutsuArgument.Companion.jutsu(context)
                    })
                )
            }

    fun register(eventBus: IEventBus) {
        ARGUMENT_TYPES.register(eventBus)
    }
}
